### DISCUSSION TITLE
[Proposal]: Can the subscription  +=  ensure that the object it is NOT already subscribed to that event ?

### DISCUSSION THREAD

**Discussion by ericwood8** [dotnet/csharplang#7332] [Language Ideas]
The problem exists - you can "over subscribe" to an event :  
https://stackoverflow.com/questions/367523/how-to-ensure-an-event-is-only-subscribed-to-once

So developers must handle this themselves by doing extra code to protect things.  This extra code gets extra ugly if they are multi-threaded and are concerned with locking.

I would like the subscriber object to  subscribe to the event (via +=  ) if there is NOT an existing event subscription of that same event.  That way developers cannot mess up.  So no change to the C# syntax, just add protections to prevent errors. 
Example existing event subscription:  _vm.PropertyChanged += Model_PropertyChanged;

**Drawbacks:**
Currently the += allows for "over subscription", so this is a change in behavior.  However, I cannot see any reason to have multiple subscriptions on the same event by the same subscriber, where the same code code would be run multiple times in response to a single event. 

CLARIFICATION - I am NOT talking about the += operation done on numbers.

**Comment by HaloFour** [comment]
This is not possible to know from the consumer's point of view, the `+=` syntax is translated to an accessor method that does not expose the underlying mechanics of the event.  It would be up to the implementation of the event accessors themselves to enforce such a policy.

As for having the compiler emit such behavior by default, while I can understand that it may be undesirable in some circumstances, I don't think one can say that it's always undesirable to the extent that it's worth potentially impacting 21 years of existing code.

**Reply by TahirAhmadov** [reply]
Yes the breaking change aspect is a complete non-starter. However, this got me thinking - I had a different question about `async` event handlers - and the real solution is a special "event" type.
```cs
// these two structs are added to a widely circulated nuget or the BCL
readonly struct Event<T> : IEnumerable<T> where T : Delegate
{
  public void Add(T handler) { ... } // thread-safety? prevent multiple additions of the same handler?
  
  readonly List<T> _handlers=new(); // what type to use here?

  IEnumerator<T> IEnumerable<T>.GetEnumerator() => throw new NotSupportedException();
}
readonly struct EventPrivate<T>
{
  public T Raise
  {
    get
    {
       // this has to be thought through
    }
  }
  public Event<T> Public => this._public;
  readonly Event<T> _public=new();
}

// then event declaration...
class Button
{
  public Event<EventHandler> Click=>this._click.Public;

  readonly EventPrivate<EventHandler> _click = new();

  void WndProc(...) { ... this._click.Raise(this, EventArgs.Empty); ... }
}

// and subscription
class MyForm
{
  public MyForm()
  {
    this.Controls.Add(new Button
    {
      Click = { this.btn_Click },
    });
  }
  void btn_Click(object s, EventArgs e) { ... }
}
```
PS. I just realized that for better or for worse, we have multi-cast delegates, so instead of `List<T>`, just `T?` is OK, albeit the structs cannot be readonly, perhaps they can be returned as `ref`?

**Reply by HaloFour** [reply]
You have a couple of choices here.  You could define the event delegate as returning `Task<T>`.  However, this will run into issues with multicasting since you'll only get one return value regardless of how many targets are invoked.  Another option would be to declare an `EventArgs` type to which the subscriber would add `Task<T>` instances, and then the publisher can enumerate/combine and await those tasks.

**Reply by TahirAhmadov** [reply]
@HaloFour yes, all that is doable - I specifically didn't go into the `async` questions - my main point was, what could an event type look like without using native .NET/C# events.
PS. The one thing that any such type cannot provide - is the signature "translation". In the above example, if the `Raise` is a property of type `TDelegate`, it sort of works for sync, but for async necessitates the creation of a hidden method which does the enumeration and awaiting, which is impossible to write once for all delegate types.

**Reply by HaloFour** [reply]
OP could theoretically be solved with source generators if C# supported `partial` events in a similar vein to [partial properties](https://github.com/dotnet/csharplang/issues/6420), where the generator emits the implementation of the event accessors to apply whatever policy is appropriate.  Identifying whether or not a subscriber is already a target in the delegate might be a bit of a pain, though, as it would probably involve recursively unwrapping the invocation list.
